src/java.base/share/classes/jdk/internal/vm/annotation/constant-folding.md

```diff
@@ -0,0 +1,128 @@
+   modification through core reflection `Field.set`.
+6. If the field is `CallSite.target`, it may be constant. (This has extra
+   treatments like nmethod dependency, so is not quite as other constant
+   fields)
+
+A `ciField` models a field declaration in the JVM, so an inherited field (as
+in a field reference, static or instance) in a subclass or subinterface shares
+the constantness settings.
+
+After a field is considered to be possibly constant, its value is fetched from
+the runtime and examined.  If the field is stable, and the value is zero or null
+(the default value), this read value is not constant.  Only non-stable final
+fields can have their zero or null values considered constant.
+
+### Array constants
+
+If an array field is stable, the type system in the compiler of Hotspot marks
+the array to be stable up to its declared level.  (See the code of the most
+generic variant of `Type::make_constant_from_field`)  As a result, access to
+nested array components _with a constant index_ can be treated as a constant
+value, if the read value is not zero or null (the default value).
+
+This means the stable annotation is not as helpful for random access (it only
+elides loading the array reference), and null components in an "immutable" array
+may cause surprising slowdowns.
+
+## How can I verify constant folding?
+
+Since constant folding makes a huge difference in API performance characteristics,
+tests are necessary to guarantee they happen.
+
+The most reliable way to ensure folding is IR tests in the compiler; we can
+expect compiler to eliminate known foldable IR structures when its inputs are
+eligible.  For example, in the initial constant folding example of `a + b`, an
+IR test can verify the int addition is eliminated in the resulting IR by
+constant folding.
+
+An example test is `compiler.c2.irTests.constantFold.TestOptional`.  Note that
+IR tests need to be run in a debug (fastdebug) configure profile, which is not
+used for most jdk library tests.
+
+JMH benchmarks can be another way to verify, except they are costly to run and
+their trend is hard to track.  Prefer the compiler tests instead.
+
+## Relation to final mechanisms
+
+### `Field.trustedFinal` property
+
+A `Field` object has a `trustedFinal` field, which when set to `true`, prevents
+core reflection or method handles from creating a setter for this field in any
+scenario.  This is derived from `fieldDescriptor::is_trusted_final()`, which
+designates final fields that are static, declared in a record class, or declared
+in a hidden class as `trustedFinal`.  This rule is different from the "may be
+constant variable" rule from above; in particular, it does not protect the
+instance final fields in eligible system classes per rule 5 above.
+
+Note that a `Field` object also models a field declaration in the JVM like a
+`ciField`, so an inherited field in a subclass or subinterface shares the
+`trustedFinal` setting.
+
+### Make Final Mean Final
```

Comment:
I think you can drop this section for now. It's okay to reference JEP 500 but it will be annoying to have to maintain this text as there are many steps to follow this one.